symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/MessageService.java

```diff
@@ -1,35 +1,118 @@
   private final AuthSession authSession;
+  private final TemplateResolver templateResolver;
+  private String template;
 
   public MessageService(MessagesApi messagesApi, AuthSession authSession) {
     this.messagesApi = messagesApi;
     this.authSession = authSession;
+    this.templateResolver = new TemplateResolver();
+  }
+
+  public MessageService(MessagesApi messagesApi, AuthSession authSession, TemplateEngine templateEngine) {
+    this.messagesApi = messagesApi;
+    this.authSession = authSession;
+    this.templateResolver = new TemplateResolver(templateEngine);
   }
 
-  public V4Message send(V4Stream stream, String message) {
+  /**
+   * Initializes the template to be used for {@link #send(V4Stream, Object)} and {@link #send(V4Stream, Object)}.
+   * Try to fetch a template from built-in templates or classpath, from the file system or from the provided URL.
+   *
+   * @param template the name, path or URL of the template to be used to send messages
+   * @return the updated {@link MessageService} instance
+   */
+  public MessageService template(String template) {
+    this.template = template;
+    return this;
+  }
+
+  /**
+   * Sends a message to the stream ID of the passed {@link V4Stream} object.
+   *
+   * @param stream the stream to send the message to
+   * @param message the message payload in MessageML
+   * @return a {@link V4Message} object containing the details of the sent message
+   * @see <a href="https://developers.symphony.com/restapi/reference#create-message-v4">Create Message v4</a>
+   */
+  public V4Message send(@Nonnull V4Stream stream, @Nonnull String message) {
     return this.send(stream.getStreamId(), message);
   }
 
-  public V4Message send(String streamId, String message) {
+  /**
+   * Sends a templated to the stream ID of the passed {@link V4Stream} object.
+   * Template must have been initialized before by calling {@link #template(String)}.
+   *
+   * @param stream the stream to send the message to
+   * @param parameters the parameters to pass to the template to produce the message to be sent
+   * @return a {@link V4Message} object containing the details of the sent message
+   * @see <a href="https://developers.symphony.com/restapi/reference#create-message-v4">Create Message v4</a>
+   */
+  public V4Message send(@Nonnull V4Stream stream, Object parameters) throws TemplateException {
+    return this.send(stream.getStreamId(), parameters);
+  }
+
+  /**
+   * Sends a templated to the stream ID passed in parameter.
+   * Template must have been initialized before by calling {@link #template(String)}.
+   *
+   * @param streamId the ID of the stream to send the message to
+   * @param parameters the parameters to pass to the template to produce the message to be sent
+   * @return a {@link V4Message} object containing the details of the sent message
+   * @throws TemplateException
+   */
+  public V4Message send(@Nonnull String streamId, Object parameters) throws TemplateException {
```

Comment:
here we couple the template().send() together but do not make it very visible
IMHO even a send(streamId, template, parameters) method would be clearer

or do we have a lot of cases when we send multiple messages using the same template?

It feels risky to have mutability as part of our services

Reply:
> It feels risky to have mutability as part of our services

+1